### DISCUSSION TITLE
Running Node-RED, running into permission problems when deploying. How do I set the userid in Dokku?

### DISCUSSION THREAD

**Discussion by cstuder** [dokku/dokku#4315] [Q&A] [answered]
How do I set the `uid/gid` to 1000 inside a Docker container in Dokku?

I've successfully launched the Docker image `nodered/node-red:1.2.7` as Dokku app, and attached a persisted volume with directory permissions 777. 

When launching Node-RED, it creates a bunch of files in the storage directory with user/group `dokku/dokku`. So far so good:

```text
drwxrwxrwx  5 root  root   4096 Jan 10 22:12 .
drwxr-xr-x  7 dokku dokku  4096 Jan 10 21:26 ..
-rw-r--r--  1 dokku dokku 14041 Jan 10 22:11 .config.nodes.json
-rw-r--r--  1 dokku dokku 14042 Jan 10 22:11 .config.nodes.json.backup
...
drwxr-xr-x  3 dokku dokku  4096 Jan 10 21:30 lib
drwxr-xr-x 46 dokku dokku  4096 Jan 10 22:02 node_modules
```

But when I want to deploy a new flow, it needs to modify files in storage and fails with the error message: `Deploy failed: Permission denied`. It looks like the problem [described here](https://discourse.nodered.org/t/node-red-deploy-failed-with-eacces-permission-denied/749).

The proposed solution is to set the `uid` within the container to 1000. How do I do that in Dokku? Do I have to write a Dockerfile?

**Comment by badsyntax** [comment, marked as answer]
I often get these permission issues with mounted directories. There's two approaches I generally take:

1. Set permissions so the files belong to nobody, or
2. Set permissions so the files belong to a specific user (your case)

In both cases, you need to set file permissions on the host, not within the container. 

For the first case, you can do something like:

```
sudo chown -R nobody:nogroup /var/lib/dokku/data/storage/app
```

For the second case (preferred), I would do something like:

```
sudo chown -R 1000:1000 /var/lib/dokku/data/storage/app
```

I would try the second approach. This also aligns with the [node-red docs](https://nodered.org/docs/getting-started/docker#using-a-host-directory-for-persistence-bind-mount):

> Note: Users migrating from version 0.20 to 1.0 will need to ensure that any existing /data directory has the correct ownership. As of 1.0 this needs to be 1000:1000. This can be forced by the command sudo chown -R 1000:1000 path/to/your/node-red/data

**Reply by nerg4l** [reply]
Note: With setuid (`chmod u+s dir`) and setgid (`chmod g+s dir`) you can also control the default group/owner in a directory.

**Reply by cstuder** [reply]
The `chown -r 1000:1000` did the trick.

I guess I don't really understand the relationship of the permissions on the host system and the permissions coming from the container. I assumed a blanked `chmod 777` would simply remove all these troubles.

Thank you very much.
